### PR TITLE
Traceable AI API Security Platform Integration

### PR DESCRIPTION
**Pull request by mtraceable**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: link to the issue

## Description
A few sentences describing the overall goals of the pull request's commits.

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Cortex XSOAR
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0
- [x] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

**Comment by michal-dagan** [comment]
Hi @mtraceabl, 
Thank you for your contribution. Your generosity and caring are unrivaled! Make sure to register your contribution by filling the [Contribution Registration](https://forms.gle/XDfxU4E61ZwEESSMA) form, so our content wizard @michal-dagan will know the proposed changes are ready to be reviewed.

**Review by michal-dagan — CHANGES_REQUESTED** [review]
Hi @mtraceable,
Thank you for your contribution!
Good Work!
1. please address my comments :)
2. please remove these files:
Pipfile
Pipfile.lock
baseintegration-dummy.json
pytest.ini
3. please revert changes in this file:
devcontainer.json
4. please change the Author_image.png according to the standard : https://xsoar.pan.dev/docs/packs/packs-format#author_imagepng
5. Please use our debugging standards instead of the logging package:
https://xsoar.pan.dev/docs/integrations/code-conventions#logging
https://xsoar.pan.dev/docs/integrations/debugging
https://xsoar.pan.dev/docs/integrations/code-conventions#return_error
https://xsoar.pan.dev/docs/reference/api/common-server-python#return_warning

Please feel free to reach out to me with any questions - I'm available here or on slack :)
Thanks again

File: Packs/Traceable/Integrations/Traceable/Traceable.py [hunk 1]
```diff
@@ -0,0 +1,666 @@
...
+      spanRequestCookies: attribute(expression: { key: "spanRequestCookies" })
+    }
+    # total
+  }
+}
+"""
+
+
+""" CLIENT CLASS """
+
```
[the remark this answers is not in the document]
Comment (mtraceable) [reply Packs/Traceable/Integrations/Traceable/Traceable.py:199]: Done

File: Packs/Traceable/Integrations/Traceable/Traceable.py [hunk 2]
```diff
@@ -0,0 +1,666 @@
...
+    base_url = demisto.params()["url"]
+
+    # if your Client class inherits from BaseClient, SSL verification is
+    # handled out of the box by it, just pass ``verify_certificate`` to
+    # the Client constructor
+    verify_certificate = not demisto.params().get("insecure", False)
+
+    # if your Client class inherits from BaseClient, system proxy is handled
+    # out of the box by it, just pass ``proxy`` to the Client constructor
+    proxy = demisto.params().get("proxy", False)
```
[the remark this answers is not in the document]
Comment (mtraceable) [reply Packs/Traceable/Integrations/Traceable/Traceable.py:605]: Done

File: Packs/Traceable/Integrations/Traceable/Traceable.py [hunk 3]
```diff
@@ -0,0 +1,666 @@
...
+        limit = int(demisto.params().get("max_fetch", 100))
+
+        _env = demisto.params().get("environment")
+
+        environments = None
+        if _env is not None and len(_env) > 0:
+            environments = []
+            _env_list = _env.split(",")
+            for _env_item in _env_list:
+                environments.append(_env_item.strip())
```
[the remark this answers is not in the document]
Comment (mtraceable) [reply Packs/Traceable/Integrations/Traceable/Traceable.py:624]: Done

File: Packs/Traceable/Integrations/Traceable/Traceable.py [hunk 4]
```diff
@@ -0,0 +1,666 @@
...
+
```
[the remark this answers is not in the document]
Comment (mtraceable) [reply Packs/Traceable/Integrations/Traceable/Traceable.py:625]: Done

File: Packs/Traceable/Integrations/Traceable/Traceable.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mtraceable) [reply Packs/Traceable/Integrations/Traceable/Traceable.py:199]: Done

File: Packs/Traceable/Integrations/Traceable/Traceable.py [hunk 5]
```diff
@@ -0,0 +1,666 @@
...
+class Helper:
+    @staticmethod
+    def construct_filterby_expression(*clauses):
+        non_null_list = [i for i in clauses if i is not None]
+        return "filterBy: [" + ",".join(non_null_list) + "]"
+
+    @staticmethod
+    def datetime_to_string(d: datetime):
+        return d.strftime("%Y-%m-%dT%H:%M:%SZ")
```
[the remark this answers is not in the document]
Comment (mtraceable) [reply Packs/Traceable/Integrations/Traceable/Traceable.py:209]: Done

File: Packs/Traceable/Integrations/Traceable/Traceable.py [hunk 6]
```diff
@@ -0,0 +1,666 @@
...
+        # "filterBy: [{keyExpression: {key: \"environment\"}, operator: IN, value: [\"" + environment + "\"], type: ATTRIBUTE}]"
+        if key is None:
+            logging.warning("Key was None. Couldn't create Key Expression.")
+            return ""
+        if operator == "IN":
+            _value = value
+            if value is not None:
+                if type(value) == str:
+                    _value = '"' + value + '"'
+                elif type(value) == int or type(value) == float:
```
[the remark this answers is not in the document]
Comment (mtraceable) [reply Packs/Traceable/Integrations/Traceable/Traceable.py:222]: Done

File: Packs/Traceable/Integrations/Traceable/Traceable.py [hunk 7]
```diff
@@ -0,0 +1,666 @@
...
+        is_error, error = self.errors_in_response(response)
+        if is_error:
+            logging.error(error)
+            raise Exception(error)
+
+        if response is not None and response.text is not None:
+            response_obj = json.loads(response.text)
+            return response_obj
+
+        raise Exception("Something went wrong: " + json.dumps(response, indent=2))
```
[the remark this answers is not in the document]
Comment (mtraceable) [reply Packs/Traceable/Integrations/Traceable/Traceable.py:346]: Done

File: Packs/Traceable/Integrations/Traceable/Traceable.py [hunk 8]
```diff
@@ -0,0 +1,666 @@
...
+    if last_fetch is None:
+        last_fetch = dateparser.parse(first_fetch_time)
+    else:
+        last_fetch = dateparser.parse(last_fetch)
+
+    latest_created_time = last_fetch
+    incidents = []
+    items = client.get_threat_events(last_fetch, datetime.now())
+    demisto.info("Retrieved " + str(len(items)) + " records.")
+    logging.debug("First Incident: " + json.dumps(items[0], indent=3))
```
[the remark this answers is not in the document]
Comment (mtraceable) [reply Packs/Traceable/Integrations/Traceable/Traceable.py:558]: Done

File: Packs/Traceable/Integrations/Traceable/Traceable.py [hunk 9]
```diff
@@ -0,0 +1,666 @@
...
+        incident_created_time = datetime.fromtimestamp(
+            item["timestamp"]["value"] / 1000
+        )
+        incident = {
+            "name": item["name"],
+            "displayname": item["displayname"],
+            "country": item["country"],
+            "sourceip": item["sourceip"],
+            "riskscore": item["riskscore"],
+            # 'severity': item['severity'],
```
[the remark this answers is not in the document]
Comment (mtraceable) [reply Packs/Traceable/Integrations/Traceable/Traceable.py:569]: Done

File: Packs/Traceable/Integrations/Traceable/Traceable.py [hunk 10]
```diff
@@ -0,0 +1,666 @@
...
+            "occurred": incident_created_time.strftime("%Y-%m-%dT%H:%M:%SZ"),
+            "rawJSON": json.dumps(item),
```
[the remark this answers is not in the document]
Comment (mtraceable) [reply Packs/Traceable/Integrations/Traceable/Traceable.py:571]: Done

**Comment by michal-dagan** [comment]
Hi @mtraceable ,
Thank you for your contribution. Your generosity and caring are unrivaled! Make sure to register your contribution by filling the [Contribution Registration](https://forms.gle/XDfxU4E61ZwEESSMA) form, so our content wizard @michal-dagan will know the proposed changes are ready to be reviewed.

File: Packs/Traceable/Integrations/Traceable/Traceable.py [hunk 11]
```diff
@@ -0,0 +1,666 @@
...
+        if operator == "IN":
+            _value = value
+            if value is not None:
+                if type(value) == str:
+                    _value = '"' + value + '"'
+                elif type(value) == int or type(value) == float:
+                    _value = value
+                elif type(value) == list and len(value) > 0:
+                    if type(value[0]) == str:
+                        _value = '"' + '","'.join(value) + '"'
```
[the remark this answers is not in the document]
Comment (mtraceable) [reply Packs/Traceable/Integrations/Traceable/Traceable.py]: This doesn't work for the use case. If the string array: CRITICAL,HIGH,MEDIUM is provided as input, the suggested change results into:
"CRITICAL,HIGH,MEDIUM,LOW"

However the required result is:
"CRITICAL", "HIGH", "MEDIUM", "LOW".

Hence have changed it to:
_value = ",".join(['"{}"'.format(v) for v in value])

File: Packs/Traceable/Integrations/Traceable/Traceable.py [hunk 12]
```diff
@@ -0,0 +1,666 @@
...
+        elif operator == "EQUALS":
+            return (
+                '{keyExpression: {key: "'
+                + key
+                + '"}, operator: '
+                + operator
+                + ", value: "
+                + value
+                + ", type: "
+                + type
```
[the remark this answers is not in the document]
Comment (mtraceable) [reply Packs/Traceable/Integrations/Traceable/Traceable.py]: Done

File: Packs/Traceable/Integrations/Traceable/Traceable.py [hunk 13]
```diff
@@ -0,0 +1,666 @@
...
+        )
+
+        if response is not None and response.status_code != 200:
+            if response.text is not None:
+                msg = (
+                    "Error occurred: "
+                    + response.text
+                    + " | Status Code: "
+                    + str(response.status_code)
+                )
```
[the remark this answers is not in the document]
Comment (mtraceable) [reply Packs/Traceable/Integrations/Traceable/Traceable.py:340]: Done

File: Packs/Traceable/Integrations/Traceable/Traceable.py [hunk 14]
```diff
@@ -0,0 +1,666 @@
...
+    def set_limit(self, limit):
+        self.limit = limit
+
+    def set_environments(self, environments):
+        self.environments = environments
+
+    def graphql_query(self, query, params={}, verify=False):
+        response = requests.post(
+            self.url,
+            json={"query": query, "variables": {}},
```
[the remark this answers is not in the document]
Comment (mtraceable) [reply Packs/Traceable/Integrations/Traceable/Traceable.py:327]: Done

File: Packs/Traceable/Integrations/Traceable/Traceable.py [hunk 15]
```diff
@@ -0,0 +1,666 @@
...
+            headers=self.headers,
+            verify=verify,
+        )
+
+        if response is not None and response.status_code != 200:
+            if response.text is not None:
```
[the remark this answers is not in the document]
Comment (mtraceable) [reply Packs/Traceable/Integrations/Traceable/Traceable.py:334]: Done

File: Packs/Traceable/Integrations/Traceable/Traceable.py [hunk 16]
```diff
@@ -0,0 +1,666 @@
...
+    def construct_key_expression(key, value, _type="ATTRIBUTE", operator="IN"):
+        # "filterBy: [{keyExpression: {key: \"environment\"}, operator: IN, value: [\"" + environment + "\"], type: ATTRIBUTE}]"
+        if key is None:
+            logging.warning("Key was None. Couldn't create Key Expression.")
+            return ""
+        if operator == "IN":
+            _value = value
+            if value is not None:
+                if type(value) == str:
+                    _value = '"' + value + '"'
```
[the remark this answers is not in the document]
Comment (mtraceable) [reply Packs/Traceable/Integrations/Traceable/Traceable.py]: Done

**Review by michal-dagan — COMMENTED** [review]
Great!! 
You can run locally: ` pre-commit -g --unit-test --validate --no-secrets --show-diff-on-failure --verbose` this will fix automatically things.
Just please revert your changes to the `devcontainer.json` file.

We're ready for a demo. Please check [this page](https://xsoar.pan.dev/docs/contributing/demo-prep), and let me know when you're available for one over DFIR.

File: Packs/Traceable/Integrations/Traceable/Traceable.py [hunk 17]
```diff
@@ -0,0 +1,687 @@
...
+s = requests.Session()
+retries = Retry(
+    total=30,
+    other=10,
+    connect=10,
+    read=10,
+    backoff_factor=1,
+    status_forcelist=[429, 500, 502, 503, 504],
+)
+s.mount("https://", HTTPAdapter(max_retries=retries))
```
[the remark this answers is not in the document]
Comment (mtraceable) [reply Packs/Traceable/Integrations/Traceable/Traceable.py:29]: The graphql service is ratelimited at Traceable end, and the ratelimiting works at the account as well as tenant level. so sometimes with no fault of ours we might get ratelimited, so this retry enables exponential backoff to circumvent the ratelimiting problem.

File: Packs/Traceable/Integrations/Traceable/Traceable.py [hunk 18]
```diff
@@ -0,0 +1,687 @@
...
+        return 4
+    elif _severity_string == "high":
+        return 3
+    elif _severity_string == "medium":
+        return 2
+    elif _severity_string == "low":
+        return 1
+    else:
+        return 0
+
```
[the remark this answers is not in the document]
Comment (mtraceable) [reply Packs/Traceable/Integrations/Traceable/Traceable.py:571]: Done

File: Packs/Traceable/Integrations/Traceable/Traceable.py [hunk 19]
```diff
@@ -0,0 +1,687 @@
...
+            item["timestamp"]["value"] / 1000
+        )
+        incident = {
+            "name": item["name"],
+            "displayname": item["displayname"],
+            "country": item["country"],
+            "sourceip": item["sourceip"],
+            "riskscore": item["riskscore"],
+            "severity": get_severity_string_to_id(item["severity"]),
+            "occurred": incident_created_time.strftime(DATE_FORMAT),
```
[the remark this answers is not in the document]
Comment (mtraceable) [reply Packs/Traceable/Integrations/Traceable/Traceable.py:598]: Done

File: Packs/Traceable/Integrations/Traceable/Traceable.py [hunk 20]
```diff
@@ -0,0 +1,687 @@
...
+        if operator == "IN":
+            _value = value
+            if value is not None:
+                if type(value) == str:
+                    _value = f'"{value}"'
+                elif isinstance(value, (int, float)):
+                    _value = value
+                elif type(value) == list and len(value) > 0:
+                    if type(value[0]) == str:
+                        _value = ",".join(['"{}"'.format(v) for v in value])
```
[the remark this answers is not in the document]
Comment (mtraceable) [reply Packs/Traceable/Integrations/Traceable/Traceable.py:217]: Done

File: Packs/Traceable/Integrations/Traceable/Traceable.py [hunk 21]
```diff
@@ -0,0 +1,687 @@
...
+                    elif type(value[0]) == int or type(value[0]) == float:
+                        _value = ",".join(["{}".format(str(v)) for v in value])
```
[the remark this answers is not in the document]
Comment (mtraceable) [reply Packs/Traceable/Integrations/Traceable/Traceable.py:219]: Done

File: Packs/Traceable/Integrations/Traceable/Traceable_test.py [hunk 22]
```diff
@@ -0,0 +1,516 @@
...
+    assert (
+        type(client.headers) == dict
+        and "Content-Type" in client.headers
+        and client.headers["Content-Type"] == "application/json"
+    )
+
+
+def test_graphql_query_non_200(mocker, caplog, capfd):
+    from Traceable import Client
+    import json
```
[the remark this answers is not in the document]
Comment (mtraceable) [reply Packs/Traceable/Integrations/Traceable/Traceable_test.py:500]: Done

File: Packs/Traceable/Integrations/Traceable/Traceable.py [hunk 23]
```diff
@@ -0,0 +1,687 @@
...
+
+class Client(BaseClient):
+    REQUESTS_TIMEOUT = 60
+
+    def __init__(
+        self,
+        base_url,
+        verify=True,
+        proxy=False,
+        ok_codes=tuple(),
```
[the remark this answers is not in the document]
Comment (mtraceable) [reply Packs/Traceable/Integrations/Traceable/Traceable.py:281]: Done

File: Packs/Traceable/Integrations/Traceable/Traceable.py [hunk 24]
```diff
@@ -0,0 +1,687 @@
...
+            verify=verify,
+        )
+        demisto.debug("Completed request...")
+
+        if (
+            response is not None
+            and response.status_code != 200
+            and response.text is not None
+        ):
+            msg = f"Error occurred: {response.text} | Status Code: {str(response.status_code)}"
```
[the remark this answers is not in the document]
Comment (mtraceable) [reply Packs/Traceable/Integrations/Traceable/Traceable.py:339]: Done

**Review by michal-dagan — CHANGES_REQUESTED** [review]
1. Please run `pre-commit -g --unit-test --validate --no-secrets --show-diff-on-failure --verbose` cause the pre-commit job failed
2. Regarding the author image - After I sent you the documentation I saw that the content is not updated, we are now working on updating it in [xsoar pan dev](https://xsoar-pan-dev--pull-request-1336-0i25g2kq.web.app/docs/packs/packs-format#author_imagepng), it should be a logo of the contributing company, you can see for example [Packs/Binalyze](https://github.com/demisto/content/tree/master/Packs/Binalyze)
3. [Tip: You should use dict.get(key) instead of dict[key]
](https://www.30secondsofcode.org/python/s/dict-getkey-vs-dictkey/#:~:text=get()%20will%20always%20return,the%20given%20key%20is%20missing.) consider to change in your code accordingly.
4. Please revert your changes to the devcontainer.json file.

thanks!!

File: Packs/Traceable/Integrations/Traceable/Traceable.py [hunk 25]
```diff
@@ -0,0 +1,679 @@
...
+                + '"}, operator: '
+                + operator
+                + ", value: ["
+                + str(_value)
+                + "], type: "
+                + _type
+                + "}"
+            )
+        elif operator == "EQUALS":
+            _value = None
```
[the remark this answers is not in the document]
Comment (mtraceable) [reply Packs/Traceable/Integrations/Traceable/Traceable.py:246]: Done

File: Packs/Traceable/Integrations/Traceable/Traceable.py [hunk 26]
```diff
@@ -0,0 +1,679 @@
...
+    ):
+        query = self.get_threat_events_query(starttime, endtime)
+        demisto.debug("Query is: " + query)
+        result = self.graphql_query(query)
+        if Helper.is_error(result, "data", "explore", "results"):
+            msg = "Error Object: " + json.dumps(result)
+            demisto.error(msg)
+            raise Exception(msg)
+
+        results = result["data"]["explore"]["results"]
```
[the remark this answers is not in the document]
Comment (mtraceable) [reply Packs/Traceable/Integrations/Traceable/Traceable.py:458]: This part actually ensures that that hierarchy exists in the result:
`if Helper.is_error(result, "data", "explore", "results"):
            msg = "Error Object: " + json.dumps(result)
            demisto.error(msg)
            raise Exception(msg)`

File: Packs/Traceable/Integrations/Traceable/Traceable.py [hunk 27]
```diff
@@ -0,0 +1,679 @@
...
+                    )
+                    continue
+                if Helper.is_error(domain_event, "spanId", "value"):
+                    demisto.info(
+                        "Couldn't find spanId in Domain Event: "
+                        + json.dumps(domain_event, indent=2)
+                    )
+                    continue
+
+                trace_id = domain_event["traceId"]["value"]
```
[the remark this answers is not in the document]
Comment (mtraceable) [reply Packs/Traceable/Integrations/Traceable/Traceable.py:481]: This is also preceeded by the call to `Helper.is_error` which ensures that the dictionary hierarchy exists, and error out if it doesn't.

File: Packs/Traceable/Integrations/Traceable/Traceable.py [hunk 28]
```diff
@@ -0,0 +1,679 @@
...
+                domain_event["severity"] = domain_event["securityScoreCategory"][
+                    "value"
+                ]
+
+            if Helper.is_error(trace_results, "data", "spans", "results"):
+                msg = "Error Object: " + json.dumps(result) + ". Couldn't get the Span."
+                demisto.info(msg)
+            else:
+                demisto.info("Found Span with id: " + span_id + ". Adding to Event.")
+                domain_event["spans"] = []
```
[the remark this answers is not in the document]
Comment (mtraceable) [reply Packs/Traceable/Integrations/Traceable/Traceable.py:532]: Done

File: Packs/Traceable/Integrations/Traceable/Traceable.yml [hunk 29]
```diff
@@ -0,0 +1,105 @@
...
+- display: Fetch incidents
+  name: isFetch
+  type: 8
+  required: false
+- display: First fetch timestamp (<number> <time unit>, e.g., 12 hours, 7 days)
+  name: first_fetch
+  defaultvalue: 1 days
+  type: 0
+  required: false
+- name: max_fetch
```
[the remark this answers is not in the document]
Comment (mtraceable) [reply Packs/Traceable/Integrations/Traceable/Traceable.yml:36]: Done

**Pull request merged**